### ISSUE TITLE
Make daml-doc work with `with` syntax

### ISSUE THREAD

**Issue by bame-da** [digital-asset/daml#1141]
This has a parse error on `t`:
```
data T = T with
  -- | test
  t : Int
```

```
parse error (possibly incorrect indentation or mismatched brackets)Compiler
```

It would be nice to be able to annotate fields in Records defined using `with` syntax.

**Comment by neil-da** [comment]
The correct syntax is:

This has a parse error on `t`:
```
data T = T with
  t : Int
    -- ^ test
```

```
parse error (possibly incorrect indentation or mismatched brackets)Compiler
```

It would be nice to be able to annotate fields in Records defined using `with` syntax.

**Comment by bame-da** [comment]
That does in deed work. Is that the intended correct syntax or the incidentally correct syntax?

**Comment by bame-da** [comment]
Haddock also supports comments before fields using `|`. Does daml-doc?
My personal preference is "comment above code". Mixing `|` and `^` style makes things rather hard to read in my opinion.

**Comment by neil-da** [comment]
Damldoc is just Haddock. If it works with `data T = T {t : Int}` but not the `with` syntax then that's a bug in `with`. Otherwise, it's a Haddock issue.

**Comment by bame-da** [comment]
OK. Only the first of the below works so we have a bug.

```
data T = T {
  -- | test
  t : Int
}

data T2 = T2 with
  -- | test
  t : Int

data T3 = T3 with
    -- | test
  t : Int
```

**Comment by neil-da** [comment]
CC @shayne-fletcher-da @rohanjr

**Comment by shayne-fletcher-da** [comment]
Yep. I suspect this is a general Haddock bug but we'll take a look.

**Comment by samuel-williams-da** [comment]
Confirmed still an issue

**Comment by samuel-williams-da** [comment]
This is a lexer issue.
When parsing, GHC placing semicolons into the generated code, unless its in `{}`
we don't use `{}` for `with`, which means `;`s are inserted. This is convenient, as it means we don't need the `,` on the cases, but it also puts those semicolons on the docs
We end up with

From:
```hs
data A = A with
  a : Int
  -- | My second annotation
  b : Int
```

gets lexed to the equivalent to
```hs
module Example where {
  data A = A with {
    a : Int;
    -- | My second annotation;
    FAILS - parser was not expecting a `;`
  }
}
```

**Comment by samuel-williams-da** [comment]
This doesn't happen in the usual flow, as these doc flags are parsed as comments when the `Opt_Haddock` flag isn't set

**Comment by samuel-williams-da** [comment]
Can't seem to find a grammer that works well for all the weird cases it currently supports + haddock, will try again another time
